src/app/components/turnos/gestor-agendas/pipes/botonesAgenda.pipe.ts

```diff
@@ -0,0 +1,125 @@
+            }
+            // Se pueden suspender agendas que estén en estado disponible o publicada...
+            if (puedeSuspender && this.puedoSuspender(agenda) && !this.tienePrestacionIniciada(agenda)) {
+                botones.suspender = true;
+            }
+            // Agregar un sobreturno
+            if (puedeDarSobreturno && this.puedoAgregar(agenda)) {
+                botones.sobreturno = true;
+            }
+            // Reasignar turnos
+            if (puedeReasignar && this.hayTurnosSuspendidos(agenda)) {
+                botones.reasignar = true;
+            }
+            // Hablita carga de pacientes masivos
+            if (puedeCargar && this.puedeCargaMasiva(agenda)) {
+                botones.cargaMasiva = true;
+            }
+            return botones;
+        }
+    }
+
+    puedoRevisar(agenda) {
+        let auditable = true;
+        for (let i = 0; i < agenda.bloques.length; i++) {
+            auditable = auditable && agenda.bloques[i].turnos.some((t: any) => t.auditable === false);
+        }
+        auditable = !auditable;
+        return (auditable && (agenda.estado === 'pendienteAsistencia' || agenda.estado === 'pendienteAuditoria' || agenda.estado === 'publicada' || agenda.estado === 'auditada') && moment(agenda.horaFin).isBefore(moment(new Date())));
+    }
+
+    puedoSuspender(agenda) {
+        return (!this.hayTurnosCodificados(agenda)) && (agenda.estado === 'disponible' || agenda.estado === 'publicada');
+    }
+
+    // Comprueba que haya algún turno con paciente, en estado suspendido
+    hayTurnosCodificados(agenda) {
+        if (!agenda.dinamica) {
+            return agenda.bloques.some((bloque: any) => bloque.turnos.some((turno: any) => turno.asistencia || (turno.diagnostico && turno.diagnostico.codificaciones && turno.diagnostico.codificaciones.length > 0)));
+        } else {
+            return agenda.bloques.some((bloque: any) => bloque.turnos.some((turno: any) => turno.asistencia));
+        }
+    }
+
+    tienePrestacionIniciada(agenda) {
+        let flag = false;
+        let b = 0;
+        while ((b < agenda.bloques.length) && (!flag)) {
+            if (agenda.bloques[b].turnos.length) {
+                const lista = agenda.bloques[b].turnos.filter(unTurno => unTurno.asistencia === 'asistio');
+                if (lista && lista.length) {
+                    flag = true;
+                }
+            }
+            b++;
+        }
+
+        return flag;
+    }
+
+    puedoAgregar(agenda) {
+        return (agenda.nominalizada && !agenda.dinamica && (agenda.estado === 'disponible' || agenda.estado === 'publicada'));
+    }
+
+    // Comprueba que haya algún turno con paciente, en estado suspendido
+    hayTurnosSuspendidos(agenda) {
+        if (!agenda.dinamica) {
+            for (let x = 0; x < agenda.bloques.length; x++) {
+                if (agenda.bloques[x].turnos) {
+                    for (let y = 0; y < agenda.bloques[x].turnos.length; y++) {
```

Comment:
preferir `forEach` en lugar de ciclos `for`. - codigo, + legible y evitás el manejo de indices